File: client-java/controller/src/main/java/org/evomaster/client/java/controller/mongo/selectors/ModSelector.java
package org.evomaster.client.java.controller.mongo.selectors;

import org.evomaster.client.java.controller.mongo.operations.*;

import java.util.List;
import java.util.Objects;

/**
 * { field: { $mod: [ divisor, remainder ] } }
 */
public class ModSelector extends SingleConditionQuerySelector {

    public static final String MOD_OPERATOR = "$mod";

    @Override
    protected QueryOperation parseValue(String fieldName, Object value) {
        Objects.requireNonNull(fieldName);
        Objects.requireNonNull(value);

        if (value instanceof List<?>) {
            List<?> listOfValues = (List<?>) value;
            if (listOfValues.size() != 2
                    || !(listOfValues.get(0) instanceof Number)
                    || !(listOfValues.get(1) instanceof Number)) {
                return null;
            }
            Long divisor = ((Number) listOfValues.get(0)).longValue();
            Long remainder = ((Number) listOfValues.get(1)).longValue();
            return new ModOperation(fieldName, divisor, remainder);
        } else {
            return null;
        }
    }

    @Override
    protected String operator() {
        return MOD_OPERATOR;
    }
}

File: client-java/controller/src/main/java/org/evomaster/client/java/controller/mongo/selectors/ModSelector.kt
package org.evomaster.client.java.controller.mongo.selectors

import org.evomaster.client.java.controller.mongo.operations.*

class ModSelector : SingleConditionQuerySelector() {
    override fun operator(): String = "\$mod"

    override fun parseValue(fieldName: String, value: Any?): QueryOperation? {
        return when (value) {
            // Validate remainder and divisor
            is List<*> -> ModOperation(fieldName, value[0] as Long, value[1] as Long)
            else -> null
        }
    }
}